src/lib/GenericPredictor.cc

```diff
@@ -23,25 +23,25 @@ GenericPredictor::GenericPredictor(ryml::ConstNodeRef config)
   btb_ =
       std::vector<std::pair<uint8_t, uint64_t>>(1 << btbBits_, {satCntVal, 0});
   // Alter globalHistoryLength_ value to better suit required format in update()
-  globalHistoryLength_ = (1 << globalHistoryLength_) - 1;
+  globalHistoryLength_ = (1 << (globalHistoryLength_ * 2)) - 1;
```

Comment:
To have a bit of an overhand so that rolling back the global history on flush does not result in 0s in the most-significant bits.  multiplying by two is a bit arbitrary, but seems to be more than sufficient.

Reply:
Yes, to roll back the history when doing speculative history updates you need to store more global history than you otherwise would.  Indeed this paper suggests exactly this solution https://jilp.org/vol2/v2paper1.pdf.  the factor of two is arbitrary.  In reality it would be as little as has empirically been shown to work, but we aren't limited by die size so I went for a very comfortable factor of two.

Reply:
If not done already I think this should be explained somewhere. Given the latter half of your comment, is this something that should be parameteriseable?

Reply:
Already done.  I don't think it makes sense make this parametrisable as it isn't the subject of much interest in the literature and a factor of two is sufficient to work.

Reply:
A user could alter this by changing the global history length I guess. But as this does impact the behaviour this should be at least noted in the docs. The docs say only n outcomes are saved when it is actually 2n for the reasons you have given. Just a line on the end of what is already there should suffice.

The multiply by 2 is not explained in the perception predictor currently so a comment should be added

Reply:
Good update. Leaving for Jack to resolve